libs/ui/src/hooks/useSmartcard.test.tsx

```diff
@@ -0,0 +1,398 @@
+  )
+}
+
+test('no card reader attached', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('no card reader')
+})
+
+test('with card reader but no card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('no card')
+})
+
+test('with card reader and a voter card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard(makeVoterCard(electionSample))
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('voter')
+})
+
+test('with card reader and a pollworker card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard(makePollWorkerCard(electionSampleDefinition.electionHash))
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('pollworker')
+})
+
+test('with card reader and an admin card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard(makeAdminCard(electionSampleDefinition.electionHash))
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('admin')
+})
+
+test('with card reader and a gibberish card', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard('not JSON')
+
+  render(<TestComponent card={card} hardware={hardware} />)
+  await advanceTimersAndPromises(CARD_POLLING_INTERVAL / 1000)
+  screen.getByText('no card')
+})
+
+test('writing short value succeeds', async () => {
+  const card = new MemoryCard()
+  const hardware = new MemoryHardware()
+
+  await hardware.setCardReaderConnected(true)
+  card.insertCard()
+
+  const TestComponent = () => {
+    const [smartcard, hasCardReader] = useSmartcard({ card, hardware })
+    const [error, setError] = useState<Error>()
+
+    useEffect(() => {
+      void (async () => {
+        const result = await smartcard?.writeShortValue(
+          JSON.stringify(makeVoterCard(electionSample))
+        )
+        setError(result?.err())
+      })()
+    }, [!smartcard])
+
+    return (
+      <div>
+        {error?.message}{' '}
+        {hasCardReader ? smartcard?.data?.t ?? 'no card' : 'no card reader'}
```

Comment:
Is all this output necessary if we're not testing for any of it? I can see how it's helpful for debugging, but — unless I'm missing something — I'd think you'd suggest removing.